### PR TITLE
Fix reverse translation of ptr.annotation for multiple UserSemantic decorations

### PR DESCRIPTION
**Pull request by vmaksimo**
It was found that the decorated variables (results of `ptr.annotation` intrinsic) are not used in further instructions unlike the input source. The issue is present for the reverse translation of multiple `UserSemantic` decorations.

This change fixes the reverse translation of multiple `ptr.annotation` intrinsics on the same class field. With this, the result of intrinsic is not lost and actually reused in further intrinsics and instructions.

### CODE REVIEW

**Review by MrSidims — COMMENTED** [review]
Overall LGTM
@tiwaria1 may I ask you to check if this patch causes regressions on your side?

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -2138,11 +2138,34 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
...
     Value *V = nullptr;
+
+    if (GEPOrUseMap.count(Base)) {
+      auto IdxToInstMap = GEPOrUseMap[Base];
+      auto Idx = AC->getIndices();
+
+      // In transIntelFPGADecorations we generated GEPs only for the fields of
+      // structure, meaning that GEP to `0` accesses the Structure itself, and
+      // the second `Id` is a Key in the map.
+      if (Idx.size() == 2) {
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply lib/SPIRV/SPIRVReader.cpp:2149]: Actually there is such case in newly added test.
```
class B {
int a <- decorated
}
class A {
 int a;
 class B b;
}
```
In this case the compiler generates GEP (0, 1) to `b` field and GEP (0,0) to the decorated member afterwards. 
At this point I can't come up with the situation where GEP with 3 indices would be generated.

File: test/transcoding/multiple_user_semantic.ll [hunk 2]
```diff
@@ -23,14 +23,19 @@
 ; CHECK-LLVM: @[[StrB:[0-9_.]+]] = {{.*}}"var_annotation_b\00"
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply test/transcoding/multiple_user_semantic.ll:23]: done

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -2138,11 +2138,34 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
...
+        }
+      }
+    }
+
     if (BB) {
       auto GEP =
           GetElementPtrInst::Create(BaseTy, Base, Index, BV->getName(), BB);
       GEP->setIsInBounds(IsInbound);
       V = GEP;
+
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply lib/SPIRV/SPIRVReader.cpp]: done

**Comment by MrSidims** [comment]
Restarting CI with a new base

**Comment by vmaksimo** [comment]
The test failure is not related to this patch. I found that the translation fails on this line `call ptr @llvm.ptr.annotation.p0.p0(ptr %16, ptr %17, ptr @.str.1, i32 12, ptr null)`, because we expect the string constant instead of bitcast as a second argument. The IR has changed after type scavenger making input incorrect.
```
// input
%11 = getelementptr inbounds %class.A, ptr %2, i32 0, i32 1 
%12 = call ptr @llvm.ptr.annotation.p0.p0(ptr %11, ptr @.str.2, ptr @.str.1, i32 12, ptr null) 
%13 = call ptr @llvm.ptr.annotation.p0.p0(ptr %12, ptr @.str.3, ptr @.str.1, i32 12, ptr null)

// IR after type scavenger magic
%15 = getelementptr inbounds %class.A, ptr %2, i32 0, i32 1 
%16 = bitcast ptr %15 to ptr 
%17 = bitcast ptr @.str.2 to ptr 
%18 = call ptr @llvm.ptr.annotation.p0.p0(ptr %16, ptr %17, ptr @.str.1, i32 12, ptr null) 
%19 = bitcast ptr @.str.3 to ptr 
%20 = call ptr @llvm.ptr.annotation.p0.p0(ptr %18, ptr %19, ptr @.str.1, i32 12, ptr null)
```

Found that #1775 fixes this issue. I believe we should merge it before this one.

**Review by MrSidims — APPROVED** [review]
LGTM
But would prefer if someone else from Intel took a look.

**Comment by vmaksimo** [comment]
@dwoodwor-intel @sarnex @asudarsa please take a look. Thanks!

**Comment by sarnex** [comment]
I don't fully understand this but if nobody else responds ping me and we can go over it

**Comment by MrSidims** [comment]
> I don't fully understand this but if nobody else responds ping me and we can go over it

That's actually a point to share some expertise and have someone with a fresh look at annotations translations, since, it does look like the initial 'design' was not the best and two authors of this 'design' are author and reviewer of this patch :D

**Comment by vmaksimo** [comment]
@MrSidims can we still merge this one? I believe no one is concerned about the change :)

**Pull request merged**